File: 205-isomorphic-strings/isomorphic-strings.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        # map: char s -> char t
        charDict = dict()
        # map: char t -> char s
        reverseDict = dict()
        for i, key in enumerate(s):
            if key not in charDict:
                if t[i] in reverseDict:
                    return False
                charDict[key] = t[i]
                reverseDict[t[i]] = key
            elif charDict[key] != t[i]:
                return False
        return True

File: 205-isomorphic-strings/isomorphic-strings.java
class Solution {
    public boolean isIsomorphic(String s, String t) {
        if(s.length() != t.length()){
            return false;
        }
        // char at index in s -> char at index in t
        HashMap<Character,Character> dict = new HashMap<>();
        // char index in t -> chart at index in s
        HashMap<Character,Character> reversedDict = new HashMap<>();


        for(int i = 0; i < s.length(); i ++){
            final char a = s.charAt(i);
            final char b = t.charAt(i);
            if(!dict.containsKey(a)){
                if(reversedDict.containsKey(b)){
                    return false;
                }
                dict.put(a,b);
                reversedDict.put(b, a);
            } else {
                if(dict.get(a) != b){
                    return false;
                } 
            }
        }

        return true;
    }
}